### PR TITLE
Clean actions, general configuration

### PR DESCRIPTION
**Pull request by ndem0**
Dear all,
I tried here to better organize all the Github actions we have at the moment:
- I merged several actions based on the trigger event
- I adjusted a bit the name of the `jobs` to improve readability
- I added the `utils/info2json.py` where there are all the explicit configurations we want to use; in this way, changing just that file, we are able to propagate changes to all the actions, without requiring manual editing anymore.

@fAndreuzzi I also merged the test for tutorials and the "classic" ones, let me know if I broke something

### CODE REVIEW

File: utils/info2json.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    info = {
+        'testing_matrix': testing_matrix_ghact,
+        'tutorial_testing_matrix': tutorial_testing_matrix_ghact,
+        'deploy_matrix': deploy_matrix_ghact,
+    }
+    parser = ArgumentParser(description='Export info using JSON output')
+    parser.add_argument('info_type', choices=list(info.keys()))
+    args = parser.parse_args()
+
+    print(json.dumps(info[args.info_type]))
```
Comment (fAndreuzzi) [inline utils/info2json.py:31]:
```suggestion
import yaml
from argparse import ArgumentParser

testing_matrix_ghact = {
    'python-version': ["3.8", "3.9", "3.10", "3.11"],
    'os': ["windows-latest", "macos-latest", "ubuntu-latest"],
    'exclude': [{'os': "windows-latest", 'python-version': "3.11"}],
}

tutorial_testing_matrix_ghact = {
    'python-version': ["3.8", "3.9", "3.10", "3.11"],
    'os': ["macos-latest", "ubuntu-latest"],
}

deploy_matrix_ghact = {
    'python-version': ["3.8"],
    'os': ["ubuntu-latest"]
}

if __name__ == '__main__':

    info = {
        'testing_matrix': testing_matrix_ghact,
        'tutorial_testing_matrix': tutorial_testing_matrix_ghact,
        'deploy_matrix': deploy_matrix_ghact,
    }
    parser = ArgumentParser(description='Export info using JSON output')
    parser.add_argument('info_type', choices=list(info.keys()))
    args = parser.parse_args()

    print(yaml.dump(info[args.info_type]))
```

File: utils/info2json.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import json
+from argparse import ArgumentParser
+
+testing_matrix_ghact = {
```
Comment (fAndreuzzi) [inline utils/info2json.py:4]: What's the meaning of `ghact`?

File: utils/info2json.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    'python-version': ["3.8", "3.9", "3.10", "3.11"],
```
Comment (fAndreuzzi) [inline utils/info2json.py:5]: Are we going to ignore `3.7`?

File: .github/workflows/testing_pr.yml [hunk 4]
```diff
@@ -6,29 +6,102 @@ on:
...
     - name: Install Python dependencies
       run: |
         pip install --upgrade pip
         pip install .[test]
-
     - name: Test with pytest
       run: |
         pytest
+
+  tutorialtest: ##############################################################################
```
Comment (fAndreuzzi) [inline .github/workflows/testing_pr.yml:65]: Same as above:
```suggestion
  tutorial_test: ##############################################################################
```

File: .github/workflows/testing_pr.yml [hunk 5]
```diff
@@ -6,29 +6,102 @@ on:
       - "master"
 
 jobs:
-  build:
+  prepare_matrix: ##############################################################################
+    runs-on: ubuntu-latest
+    outputs:
+       matrix_unittest: ${{ steps.setmatrix.outputs.matrix1 }}
+       matrix_tutorialtest: ${{ steps.setmatrix.outputs.matrix2 }}
```
Comment (fAndreuzzi) [inline .github/workflows/testing_pr.yml:13]:
```suggestion
       matrix_unit_test: ${{ steps.setmatrix.outputs.matrix1 }}
       matrix_tutorial_test: ${{ steps.setmatrix.outputs.matrix2 }}
```

File: .github/workflows/testing_pr.yml [hunk 6]
```diff
@@ -6,29 +6,102 @@ on:
...
+        run: |
+          sudo npm install -g json2yaml
+      - name: Check matrix definition
+        run: |
+          matrix='${{ needs.prepare_matrix.outputs.matrix }}'
+          echo $matrix
+          echo $matrix | jq .
+          echo $matrix | json2yaml
+
+  unittest: ##############################################################################
```
Comment (fAndreuzzi) [inline .github/workflows/testing_pr.yml:43]: I propose to use the underscore like in `check_matrix`:
```suggestion
  unit_test: ##############################################################################
```

File: .github/workflows/testing_pr.yml [hunk 7]
```diff
@@ -6,29 +6,102 @@ on:
...
+    needs: prepare_matrix
     runs-on: ${{ matrix.os }}
-    strategy:
-      fail-fast: false
-      matrix:
-        os: [windows-latest, macos-latest, ubuntu-latest]
-        python-version: ["3.7", "3.8", "3.9", "3.10"]
+    strategy: 
+      matrix: ${{ fromJson(needs.prepare_matrix.outputs.matrix_unittest) }}
```
Comment (fAndreuzzi) [inline .github/workflows/testing_pr.yml:47]:
```suggestion
      matrix: ${{ fromJson(needs.prepare_matrix.outputs.matrix_unit_test) }}
```

File: .github/workflows/testing_pr.yml [hunk 8]
```diff
@@ -6,29 +6,102 @@ on:
...
-
     - name: Test with pytest
       run: |
         pytest
+
+  tutorialtest: ##############################################################################
+    needs: prepare_matrix
+    runs-on: ${{ matrix.os }}
+    strategy: 
+      matrix: ${{ fromJson(needs.prepare_matrix.outputs.matrix_tutorialtest) }}
```
Comment (fAndreuzzi) [inline .github/workflows/testing_pr.yml:69]:
```suggestion
      matrix: ${{ fromJson(needs.prepare_matrix.outputs.matrix_tutorial_test) }}
```

File: utils/info2json.py (hunk 2 above)
Comment (ndem0) [reply utils/info2json.py:4]: github action, I just keep an open strucutre to export in future other information as json format (e.g. version)

File: utils/info2json.py (hunk 3 above)
Comment (ndem0) [reply utils/info2json.py:5]: It was just a try, for the moment I would keep the 3.7

**Pull request merged**